server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1297,10 +1295,9 @@ public ClusterState execute(ClusterState currentState) {
                     request.indexSettings(),
                     request.ignoreIndexSettings()
                 );
-                if (snapshotIndexMetadata.getCreationVersion()
-                    .before(currentState.getNodes().getMaxNodeVersion().minimumIndexCompatibilityVersion())) {
+                if (snapshotIndexMetadata.getCompatibilityVersion().before(minIndexCompatibilityVersion)) {
```

Comment:
For the "restore/mount from snapshot" flow, I'm not sure how someone would "inadvertently" restore/mount an old index. They have chosen the index to be restored, so it should be restored (today such a restore would fail, in the future it will go through). It might help if we would show index version information via snapshot status APIs so that users know in advance which indices might be older and which ones not so that they can do a more informed choice on restore. This feels like a problem to be solved on the snapshot discovery APIs (see backlinks to https://github.com/elastic/elasticsearch/issues/54940). WDYT?

For the major version upgrade of a cluster, there are a number of paths to explore. I'm less focusing right now on the second scenario you described, as that only becomes relevant once we are at the end of 8.x (as there is no 7.x releases of this feature), i.e. the only approach we can recommend in the immediate future is to allow users to take a snapshot in their old cluster, and restore that to a parallel 8.x cluster to try it out. What you're describing will certainly be useful once we go towards the next major. What I'm also thinking more deeply about right now is whether we can find a way to allow users to do a rolling / full cluster restart upgrade from 7.last to 8.x with 6.x indices still present in the cluster that satisfy certain conditions (e.g. warm/cold/frozen tiers) and that will be auto-upgraded to legacy mode, not having to trigger the process via snapshots. Here, we could also require users to take certain steps using existing APIs on 7.x (e.g. flush/mark index read-only) for this auto-upgrade to be successful.

Reply:
> I'm not sure how someone would "inadvertently" restore/mount an old index

One case is to do a restore of all indices. If you restore a 7.x snapshot in an 8.x one, I think we would silently degrade functionality and performance of 6.x indices. My intuition would be to error instead and then ask to add an explicit "archive everything before 7.0" mode in the request.

> upgrade from 7.last to 8.x with 6.x indices still present in the cluster that satisfy certain conditions

I think our upgrade assistant would advice to reindex such data. I wonder if we would not need to extend that with this functionality - and thus let it explicitly set a flag to archive on upgrade (with a warning to upgrade to at least 8.some-version)?

Reply:
> One case is to do a restore of all indices. If you restore a 7.x snapshot in an 8.x one, I think we would silently degrade functionality and performance of 6.x indices. My intuition would be to error instead and then ask to add an explicit "archive everything before 7.0" mode in the request.

I'm not sure about the situation where you would encounter this (restore of all indices into next major):
- Disaster recovery into a newer major version cluster than what you had when the snapshot was taken?
- Some kind of upgrade process where the upgrade happens via snapshot?

I'm generally wary of having to require an extra flag, as it means that any existing place / UI that allows operating with / restoring snapshots will have to become aware of this flag. Not requiring an extra flag would make this new capability of restoring older indices directly available everywhere.

For the in-place upgrade (upgrading cluster to next major version, with some indices in the cluster now becoming legacy indices), I agree that this should not just happen silently.

Reply:
OK, thanks for your thoughts on this, I do agree there is a trade-off here.